Utils/ModGen/modgen.1
.TH MODGEN 1
.SH NAME
modgen \- creates a LUTEar library module

.SH SYNOPSIS
.B modgen
.I file.mod
.br
.B modgen

.SH DESCRIPTION
The argument
.I file.mod
contains the specification of a module to be included in the
.I LUTEar
library.
If the
.I
modgen
command is issued without an argument, the user will be prompted for
.I file.mod.
Using the specification contained within
.I file.mod,
.I modgen
creates the
.I file.c
source and
.I file.h
header files required for adding a module to the
.I LUTEar
library.

.PP
The layout of the module file,
.I file.mod,
is very similar to a standard ANSI C structure typedef declaration, with the
addition of a few keywords, i.e.

.RS 10
.PP
.B mod_name		
.I moduleName;
.br
.B qualifier			
.I qualifier;
.br
.B proc_routine	
.I ProcessRoutineName;
.br
.B typedef struct
{

.I	... Parameter declarations ...

}
.I StructureName;
.RE

- keywords are shown in bold-face and the identifers, or names, are given as
bold italics. The standard ANSI C type specifiers are available for declaring
the modul parameters, e.g. 'char', 'int', 'float', 'double', and so on. Here is
an example showing the module specification for the StMPTone module.

.PP
.RS 5
/* This is a comment */
.br
.B mod_name		
PureTone;
.br
.B qualifier			
Multi;
.br
.B proc_routine	
GenerateSignal;
.br
.B typedef struct
{

.B	double	
duration, dt;
.br
.B	int_al		
numPTones;
.br
.B	double	
*intensities, *frequencies, *phases;

} MPureTone;

.SS Keywords

Here a description is given of the special keywords used by the
.I modgen
program.

.TP 9
.B mod_name
specifies the base name of the module.  This must be declared, as it appears
as the base name for all the module routines, separated from the function name
by an underscore, e.g.

PrintPars_
.I ModuleName
(
.B void
).


.TP 10
.B qualifier
specifies an additional qualifier for the module name.  This declaration is
optional, but if declared it will appear after the base name of the module in
all module routines, separated by an underscore, e.g.

PrintPars_
.I ModuleName
_
.I Qualifier
(
.B void
).

.TP 13
.B proc_routine
specifies the name to be given to the main process routine, i.e. the routine
which actually generates the output signal.  It is is not declared then the
default name

Process_
.I ModuleName
_
.I Qualifier
(...)

will be given.

.TP 7
.B int_al
is a special form of the standard 'int' type.  The '_al' stands for 'array
limit', as it is used to indicate that the respective parameter is to be used
as the array length for any following array declarations (see the example for
the StMPTone module specification given above).  Only one parameter will be
used for
.I int_al
declaration lines, and it will be applied as the array limit for all pointers
thereafter declared until the next
.I int_al
declaration.

.SH DIAGNOSTICS

If an error is encountered in the
.I file.mod
module specification file the program will exit, try to indicating the
approximate position of the error.

.SH BUGS
When declaring module parameters combinations such as 'unsigned int' will not
be interpreted correctly.  This problem can be circumvented by introducing a
typedef before the main module structure declaration, e.g.

.PP
.RS 15
.B typedef unsigned int
.I uint
.RE

.PP
This version (1.2) of the program does not know how to treat 'private' process
variables.

.SH AUTHOR
Lowel P. O'Mard
.I L.P.OMard@lut.ac.uk


DSAMSrc/ChangeLog
date: 2000/10/12 16:50:55;  author: lowel;  state: Exp;  lines: +1 -1
The 'PROCESS_START_TIME' should be set to zero and not 1.  I have changed all
files to correspond with this notion of 'instantaneous time'.



Utils/ModGen/ModGen.c
/**********************
 *
 * File:		ModuleProducer.c
 * Purpose: 	Module generator main process routines.
 * Comments:	
 * Author:		L.P.O'Mard
 * Created:		13 Nov 1995
 * Updated:	
 * Copyright:	(c) 1995, Loughborough University of Technology
 *
 **********************/

#include <stdio.h>
#include <ctype.h>

#include "MGGeneral.h"
#include "Symbol.h"
#include "Init.h"
#include "Token.h"
#include "Parser.h"
#include "WriteCodeUtils.h"
#include "WriteHeaderCode.h"
#include "WriteSrcCode.h"
#include "Strings.h"
#include "ModGen.h"

/******************************************************************************/
/****************************** Gobal variables *******************************/
/******************************************************************************/

/******************************************************************************/
/****************************** Function Declarations *************************/
/******************************************************************************/

/****************************** TestProcess ***********************************/

/* 
 * This routine was written to test this application.
 */

void
TestProcess(char *baseName)
{
	static char	*moduleName = "PureTone", *qualifier = "Multi";
	static char	*structure = "MPureTone", *headerFile = "StMPTone.h";
	int		i;

	PrintTokens(pc);
	PrintFileHeader(stdout, headerFile);
	PrintCommentHeading(stdout, "Constant definitions");
	PrintCommentHeading(stdout, "Type definitions");
	PrintExpandedStructure(stdout);
	PrintCommentHeading(stdout, "External variables");
	PrintExternalVariables(stdout);
	PrintCommentHeading(stdout, "Function Prototypes");
	Print(stdout, "\t", "A short long line which should be over eighty "
	  "characters long.  At least it will be if I keep typing.");
	Print(stdout, "\t", "A short long line which should be over eighty "
	  "characters long.  At least it will be if I keep typing.");
	Print(stdout, "", "");
	printf("\n");
	Print(stdout, "\t", "\"A short long line which should be over eighty "
	  "characters long.  At least it will be if I keep typing.\"");
	Print(stdout, "", "");
	printf("\n");
	Print(stdout, "\t", "\"A short long line which should be over "
	  "eighty characters long.  At least it _\"will_\" be if I keep typing.\"");
	Print(stdout, "", "");
	printf("\n");
	Print(stdout, "\t", "\"A short long line which should be_\"\n_\""
	  "over eighty characters long.  At least it will be if I keep typing.\"");
	Print(stdout, "", "");
	printf("\n");
	Print(stdout, "\t\t  ", "BOOLN	SetPars_PureTone_Multi(int "
	  "theNumPTones, double *theFrequencies, double *theIntensities, double "
	  "*thePhases, double theDuration, double theSamplingInterval);");
	Print(stdout, "", "");
	printf("\n");
	PrintIncludeFiles(stdout, headerFile);
	PrintGlobalVariables(stdout);
	for (i = 0; standardFunctions[i].name != NULL; i++)
		(*standardFunctions[i].func)(stdout);
	PrintFuncPrototypes(stdout);

}

/****************************** Process ***************************************/

/* 
 * This routine was written to test this application.
 * If a qualifier is found, then the name of the next token is used.
 * This is because the qualifier might have the same name as the typedef
 * structure, and hence will have been re-typed to TYPEDEF_NAME and will
 * not be found using GetType_IdentifierList().
 * The above also applies to the MOD_NAME.  This is not a good method, and
 * should be reviewed.
 */

void
Process(char *baseName)
{
	char	headerFileNamePath[MAX_FILE_PATH], mainFileNamePath[MAX_FILE_PATH];
	char	*p, *up, *headerFileName, *mainFileName;
	char	upperCaseBaseName[MAXLINE];
	int		i;
	FILE	*headerFP, *mainFP;

	snprintf(headerFileNamePath, MAX_FILE_PATH, "%s.h", baseName);
	if ((headerFP = fopen(headerFileNamePath, "w")) == NULL)
		execerror("ModGen: Process: Cannot open file", headerFileName);
	headerFileName = GetFileNameFPath(headerFileNamePath);
	snprintf(mainFileNamePath, MAX_FILE_PATH, "%s.c", baseName);
	if ((mainFP = fopen(mainFileNamePath, "w")) == NULL)
		execerror("ModGen: Process: Cannot open file", mainFileName);
	mainFileName = GetFileNameFPath(mainFileNamePath);
	
	Init_WriteCodeUtils();

	/* Main file. */
	PrintFileHeader(mainFP, mainFileName);
	PrintIncludeFiles(mainFP, headerFileName);
	PrintCommentHeading(mainFP, "Global variables");
	PrintGlobalVariables(mainFP);
	PrintCommentHeading(mainFP, "Subroutines and functions");
	for (i = 0; standardFunctions[i].name != NULL; i++)
		(*standardFunctions[i].func)(mainFP);
	fclose(mainFP);

	/* Header file. */
	
	PrintFileHeader(headerFP, headerFileName);
	for (p = GetFileNameFPath(baseName), up = upperCaseBaseName; *p != '\0';
	  p++, up++)
		*up = toupper(*p);
	*up = '\0';
	fprintf(headerFP, "#ifndef _%s_H\n", upperCaseBaseName);
	fprintf(headerFP, "#define _%s_H 1\n\n", upperCaseBaseName);
	PrintHeaderIncludes(headerFP);
	PrintCommentHeading(headerFP, "Constant definitions");
	PrintConstantDefinitions(headerFP);
	PrintCommentHeading(headerFP, "Type definitions");
	PrintParSpecifierEnumDefinition(headerFP);
	PrintNameSpecifierEnumDefinition(headerFP);
	PrintExpandedStructure(headerFP);
	PrintCommentHeading(headerFP, "External variables");
	PrintExternalVariables(headerFP);
	PrintCommentHeading(headerFP, "Function Prototypes");
	PrintFuncPrototypes(headerFP);
	fprintf(headerFP, "#endif\n");
	fclose(headerFP);

	printf("Module files written to '%s.c' and '%s.h'.\n", baseName, baseName);
	
}

/******************************************************************************/
/****************************** Main function *********************************/
/******************************************************************************/

int
main(int argc, char * argv[])
{
	int		yyparse(void);
	char	*baseName, fileName[MAXLINE];

	progname = argv[0];
	setjmp(begin);
	signal(SIGFPE, fpecatch);
	Init();
	if (argc >= 2)
		sprintf(fileName, "%s", argv[1]);
	else {
		printf("Input module file name: ");
		scanf("%s", fileName);
	}
	if ((baseName = StripFileExtension(fileName, "mod")) == NULL) {
		fprintf(stderr, "ModuleProducer: file must have '.mod' extension.\n");
		return 1;
	}
	if ((fpIn = fopen(fileName, "r")) == NULL) {
		fprintf(stderr, "ModuleProducer: could not open module file\n",
		  fileName);
		 return 1;
	}
	if (yyparse() == 1) {
		fprintf(stderr, "ModuleProducer: exit with error\n", fileName);
		return 1;
	}
	fclose(fpIn);
	TransposeTokenTypes();
	printf("%s version %s (%s):\n", progname, MODGEN_VERSION, __DATE__);
	Process(baseName);
	return 0;

}

